### PR TITLE
[DB-473] [risk=low] Unrolling snomed measurements

### PR DESCRIPTION
**Pull request by SrushtiGangireddy**
1. Removing the queries which were rolling up stratum specific counts.
2. Changed the cdr config in public, stable and staging config files.
3. Added full text index on criteria code and name columns to match them against the keywords to fetch drug ingredients (as like was fetching irrelevant data).
4. Updated the way measurement queries were binning data (After observing that o unit measurements which also have unit source value null were not being counted at all).

### CODE REVIEW

File: public-api/db-cdr/generate-cdr/generate_criteria_stratum.sh [hunk 1]
```diff
@@ -44,166 +44,6 @@ then
   exit 1
 fi
 
-echo "creating criteria_stratum table rows for snomed measurements"
```
Comment (SrushtiGangireddy) [inline public-api/db-cdr/generate-cdr/generate_criteria_stratum.sh:47]: Removed the queries that were generating rolled up counts of each stratum for snomed measurements.

File: public-api/db-cdr/generate-cdr/make-bq-data.sh [hunk 2]
```diff
@@ -628,14 +628,6 @@ from (select concept_id, est_count from \`$OUTPUT_PROJECT.$OUTPUT_DATASET.criter
 where cr.concept_id = concept_id and cr.type='SNOMED' and cr.subtype='PCS' and cr.synonyms like '%rank1%') as sub_cr
 where sub_cr.concept_id = c.concept_id and c.domain_id = 'Procedure' and c.vocabulary_id = 'SNOMED' "
 
-echo "Updating rolled up counts of snomed measurements in concept"
```
Comment (SrushtiGangireddy) [inline public-api/db-cdr/generate-cdr/make-bq-data.sh:631]: Not copying main rolled up count from concept for snomed measurements.

File: public-api/db-cdr/generate-cdr/run-achilles-queries.sh [hunk 3]
```diff
@@ -1934,11 +1934,11 @@ bq --quiet --project=$BQ_PROJECT query --nouse_legacy_sql \
 with pm_concepts as
 (select measurement_concept_id as concept,co.person_id as person
 from \`${BQ_PROJECT}.${BQ_DATASET}.measurement\` co join \`${BQ_PROJECT}.${BQ_DATASET}.concept\` c on co.measurement_concept_id=c.concept_id
-where c.vocabulary_id != 'PPI' and c.concept_id not in (3036277,903118,903115,3025315,903135,903136,903126,903111,42528957)),
+where c.concept_id in (3036277, 3025315, 3027018, 3031203, 40759207, 903107, 903126, 40765148, 903135, 903136, 3022318, 3012888, 3004249, 903115, 903118, 3038553)),
```
Comment (SrushtiGangireddy) [inline public-api/db-cdr/generate-cdr/run-achilles-queries.sh:1937]: Getting pm participant count.

File: public-api/db-cdr/generate-cdr/run-measurement-queries.sh [hunk 4]
```diff
@@ -204,359 +204,21 @@ FROM  priorstats p
 join overallstats o on p.stratum1_id = o.stratum1_id and p.stratum2_id = o.stratum2_id
 group by o.stratum1_id, o.stratum2_id, o.total, o.min_value, o.max_value, o.avg_value, o.stdev_value"
 
-# 1816 Measurement response by age at occurrence distribution
```
Comment (SrushtiGangireddy) [inline public-api/db-cdr/generate-cdr/run-measurement-queries.sh:207]: This data is not being used at this point of time. So removing the queries, saved on local system for reference.

File: public-api/db-cdr/generate-cdr/run-measurement-queries.sh [hunk 5]
```diff
@@ -611,9 +273,9 @@ count(distinct p1.person_id) as source_count_value
 from \`${BQ_PROJECT}.${BQ_DATASET}.measurement\` m1 join \`${BQ_PROJECT}.${BQ_DATASET}.person\` p1 on p1.person_id = m1.person_id
 join measurement_quartile_data on m1.measurement_concept_id=concept
 join \`${BQ_PROJECT}.${BQ_DATASET}.concept\` c1 on m1.measurement_concept_id=c1.concept_id
-join \`${WORKBENCH_PROJECT}.${WORKBENCH_DATASET}.unit_map\` um on m1.unit_source_value=um.unit_source_value
 where m1.measurement_concept_id > 0
-and m1.value_as_number is not null and p1.gender_concept_id=gender and (cast(m1.unit_concept_id as string)=unit or cast(um.unit_concept_id as string)=unit)
+and m1.value_as_number is not null and p1.gender_concept_id=gender and cast(m1.unit_concept_id as string)=unit
+and m1.unit_concept_id != 0
```
Comment (SrushtiGangireddy) [inline public-api/db-cdr/generate-cdr/run-measurement-queries.sh:279]: Binning the measurements that have unit_concept_id.

File: public-api/db-cdr/generate-cdr/run-measurement-queries.sh [hunk 6]
```diff
@@ -655,344 +317,256 @@ round((case when m1.value_as_number < p10_value then p10_value
...
 join \`${BQ_PROJECT}.${BQ_DATASET}.concept\` c1 on m1.measurement_concept_id=c1.concept_id
-join \`${WORKBENCH_PROJECT}.${WORKBENCH_DATASET}.unit_map\` um on m1.unit_source_value=um.unit_source_value
+join \`${WORKBENCH_PROJECT}.${WORKBENCH_DATASET}.unit_map\` um on lower(m1.unit_source_value)=lower(um.unit_source_value)
 where m1.measurement_concept_id > 0
-and m1.value_as_number is not null
-and floor((extract(year from m1.measurement_date) - p1.year_of_birth)/10)=ar.age_decile
-and (cast(m1.unit_concept_id as string)=unit or cast(um.unit_concept_id as string)=unit)
-and (extract(year from m1.measurement_date) - p1.year_of_birth) >= 18 and (extract(year from m1.measurement_date) - p1.year_of_birth) < 30
+and m1.value_as_number is not null and p1.gender_concept_id=gender and cast(um.unit_concept_id as string)=unit
+and m1.unit_concept_id = 0 and m1.unit_source_value is not null
```
Comment (SrushtiGangireddy) [inline public-api/db-cdr/generate-cdr/run-measurement-queries.sh:384]: Binning measurements that do not have unit concept id by joining unit_source_value on the manual map made.

File: public-api/db-cdr/generate-cdr/run-measurement-queries.sh [hunk 7]
```diff
@@ -655,344 +317,256 @@ round((case when m1.value_as_number < p10_value then p10_value
...
+from \`${BQ_PROJECT}.${BQ_DATASET}.measurement\` m1 join \`${BQ_PROJECT}.${BQ_DATASET}.person\` p1 on p1.person_id = m1.person_id
+join measurement_quartile_data on m1.measurement_concept_id=concept and unit='0'
 join \`${BQ_PROJECT}.${BQ_DATASET}.concept\` c1 on m1.measurement_concept_id=c1.concept_id
-where m1.value_as_concept_id != 0
-and m1.measurement_concept_id > 0
-and floor((extract(year from m1.measurement_date) - p1.year_of_birth)/10) >=3
-group by stratum_1,stratum_3,stratum_4,stratum_5
+where m1.measurement_concept_id > 0
+and m1.value_as_number is not null and p1.gender_concept_id=gender
+and m1.unit_concept_id = 0 and m1.unit_source_value is null
```
Comment (SrushtiGangireddy) [inline public-api/db-cdr/generate-cdr/run-measurement-queries.sh:488]: Binning all the non-unit measurements together with 'no unit' as unit name.

File: public-api/db-cdr/generate-cdr/run-measurement-queries.sh [hunk 8]
```diff
@@ -655,344 +317,256 @@ round((case when m1.value_as_number < p10_value then p10_value
...
-count(distinct p1.person_id) as source_count_value
-FROM \`${BQ_PROJECT}.${BQ_DATASET}.measurement\` m1
-join \`${BQ_PROJECT}.${BQ_DATASET}.person\` p1 on p1.person_id = m1.person_id
-join \`${BQ_PROJECT}.${BQ_DATASET}.concept\` c2 on c2.concept_id=m1.value_as_concept_id
-where m1.value_as_concept_id != 0
-and (extract(year from m1.measurement_date) - p1.year_of_birth) >= 18 and (extract(year from m1.measurement_date) - p1.year_of_birth) < 30
-and m1.measurement_source_concept_id > 0 and
-m1.measurement_source_concept_id not in (select distinct measurement_concept_id from \`${BQ_PROJECT}.${BQ_DATASET}.measurement\`)
-group by stratum_1,stratum_3,stratum_4,stratum_5"
+ (id, analysis_id, stratum_1,stratum_2,stratum_3,stratum_4,stratum_5,count_value,source_count_value)
```
Comment (SrushtiGangireddy) [inline public-api/db-cdr/generate-cdr/run-measurement-queries.sh:544]: Just got the indentation difference i guess did not make any change.

File: public-api/src/main/java/org/pmiops/workbench/publicapi/DataBrowserController.java [hunk 9]
```diff
@@ -101,10 +101,8 @@
     public static final long MEASUREMENT_DIST_ANALYSIS_ID = 1815;
 
     public static final long MEASUREMENT_GENDER_DIST_ANALYSIS_ID = 1815;
-    public static final long MEASUREMENT_AGE_DIST_ANALYSIS_ID = 1816;
```
Comment (SrushtiGangireddy) [inline public-api/src/main/java/org/pmiops/workbench/publicapi/DataBrowserController.java:104]: Not being used yet.

File: public-api/src/main/java/org/pmiops/workbench/publicapi/DataBrowserController.java [hunk 10]
```diff
@@ -101,10 +101,8 @@
...
 
     public static final long MEASUREMENT_GENDER_ANALYSIS_ID = 1900;
-    public static final long MEASUREMENT_AGE_ANALYSIS_ID = 1901;
```
Comment (SrushtiGangireddy) [inline public-api/src/main/java/org/pmiops/workbench/publicapi/DataBrowserController.java:107]: Not being used yet.

File: public-api/src/main/java/org/pmiops/workbench/publicapi/DataBrowserController.java [hunk 11]
```diff
@@ -813,30 +810,6 @@ public SurveyQuestionResult apply(AchillesResult o) {
                     addGenderStratum(aa,3);
                     isMeasurement = true;
                     conceptAnalysis.setMeasurementValueGenderAnalysis(unitSeperateAnalysis.stream().map(TO_CLIENT_ANALYSIS).collect(Collectors.toList()));
-                }else if(analysisId == MEASUREMENT_AGE_ANALYSIS_ID){
```
Comment (SrushtiGangireddy) [inline public-api/src/main/java/org/pmiops/workbench/publicapi/DataBrowserController.java:816]: Not being used anywhere yet.

File: common-api/src/main/java/org/pmiops/workbench/cdr/dao/ConceptDao.java [hunk 12]
```diff
@@ -73,8 +73,7 @@
             "where c.type = 'DRUG' " +
             "and c.subtype in ('BRAND') " +
             "and c.is_selectable = 1 " +
-            "and (upper(c.name) like upper(concat('%',?1,'%')) " +
-            "or upper(c.code) like upper(concat('%',?1,'%'))) " +
+            "and match(c.name, c.code) against(?1 in boolean mode) > 0 " +
```
Comment (SrushtiGangireddy) [inline common-api/src/main/java/org/pmiops/workbench/cdr/dao/ConceptDao.java:76]: "like" fetches un-relevant results.
Like this call fetches the ingredients of drug "diabetesee" when searched with the keyword "diabetes".
Changed this to match after testing few queries.

Comment (calbach) [reply common-api/src/main/java/org/pmiops/workbench/cdr/dao/ConceptDao.java:76]: I can't tell at a glance whether MySQL will optimize this query properly. Once this is in the test env, please check the request latency here (alternatively, you can execute this SQL directly against the test DB to verify the latency). In the other case that I optimized the SQL query alone was taking ~8s.

File: public-api/db-cdr/generate-cdr/run-measurement-queries.sh [hunk 13]
```diff
@@ -655,344 +317,256 @@ round((case when m1.value_as_number < p10_value then p10_value
...
-count(distinct p1.person_id) as source_count_value
-FROM \`${BQ_PROJECT}.${BQ_DATASET}.measurement\` m1
-join \`${BQ_PROJECT}.${BQ_DATASET}.person\` p1 on p1.person_id = m1.person_id
-join \`${BQ_PROJECT}.${BQ_DATASET}.concept\` c2 on c2.concept_id=m1.value_as_concept_id
-where m1.value_as_concept_id != 0
-and floor((extract(year from m1.measurement_date) - p1.year_of_birth)/10) >=3
-and m1.measurement_source_concept_id > 0 and
-m1.measurement_source_concept_id not in (select distinct measurement_concept_id from \`${BQ_PROJECT}.${BQ_DATASET}.measurement\`)
-group by stratum_1,stratum_3,stratum_4,stratum_5"
+select 0 as id, 1900 as analysis_id,
```
Comment (calbach) [inline public-api/db-cdr/generate-cdr/run-measurement-queries.sh:491]: Could you put a comment for this section? At a glance it looks identical to the above, what is different about this second query?

**Review by calbach — APPROVED** [review]

File: common-api/src/main/java/org/pmiops/workbench/cdr/dao/ConceptDao.java (hunk 12 above)
Comment (SrushtiGangireddy) [reply common-api/src/main/java/org/pmiops/workbench/cdr/dao/ConceptDao.java:76]: I just searched with tylenol in sequel pro 
"select distinct c.concept_id from criteria c
where c.type = 'DRUG'
and c.subtype in ('BRAND') 
and c.is_selectable = 1 
and match(c.name,c.code) against('tylenol' in boolean mode);" ~ 39.7 ms
"select distinct c.concept_id from criteria c
where c.type = 'DRUG'
and c.subtype in ('BRAND') 
and c.is_selectable = 1 
and (upper(c.name) like upper(concat('%','tylenol','%')) or upper(c.code) like upper(concat('%','tylenol','%')));" ~ 254ms

Comment (calbach) [reply common-api/src/main/java/org/pmiops/workbench/cdr/dao/ConceptDao.java:76]: Against test cloud SQL right? Not local? If so, that's pretty slow but probably acceptable if this is not a super common endpoint.

File: public-api/db-cdr/generate-cdr/run-measurement-queries.sh (hunk 13 above)
Comment (SrushtiGangireddy) [reply public-api/db-cdr/generate-cdr/run-measurement-queries.sh:491]: Added.

File: common-api/src/main/java/org/pmiops/workbench/cdr/dao/ConceptDao.java (hunk 12 above)
Comment (SrushtiGangireddy) [reply common-api/src/main/java/org/pmiops/workbench/cdr/dao/ConceptDao.java:76]: This is against the test cloudsql. I would not say its uncommon. Each time the concept search is hit, this method is called to fetch ingredients of any matching drug brands.
Match seems to work better over like or like for few queries i have tested.
Do we have a better and quick option that could speed up things ?

Comment (calbach) [reply common-api/src/main/java/org/pmiops/workbench/cdr/dao/ConceptDao.java:76]: What was the runtime for the previous `LIKE` approach?

I think this is probably fine, I can rerun the load tests afterwards to check.

Comment (SrushtiGangireddy) [reply common-api/src/main/java/org/pmiops/workbench/cdr/dao/ConceptDao.java:76]: previous like took 254 ms, match took 39.7 ms only for tylenol search
like ~ 215ms, match 35.6 ms for "percocet" search

Comment (calbach) [reply common-api/src/main/java/org/pmiops/workbench/cdr/dao/ConceptDao.java:76]: Oh - great. Nice improvement then.

**Review by calbach — APPROVED** [review]

**Comment by SrushtiGangireddy** [comment]
Thanks merging in.

**Pull request merged**